libmscore/figuredbass.cpp

```diff
@@ -1770,6 +1773,256 @@ FiguredBass* Score::addFiguredBass()
+      
+void FiguredBass::realizeFiguredBass(Score *score) {
+      Q_ASSERT(score);
 
+      for (Measure* measure = score->firstMeasure(); measure; measure = measure->nextMeasure()) {
+            realizeFiguredBass(measure);
+      }
+}
+      
+void FiguredBass::realizeFiguredBass(Measure *measure) {
+      Q_ASSERT(measure);
+      const Segment::Type st = Segment::Type::ChordRest;
+      for (Segment* segment = measure->first(st); segment; segment = segment->next(st)) {
+            realizeFiguredBass(segment);
+      }
+}
+      
+void FiguredBass::realizeFiguredBass(Segment *segment) {
+      Q_ASSERT(segment);
+      int etrack = segment->score()->nstaves() * VOICES;
+
+      for (int track = 0; track < etrack; ++track) {
+            Element *element = segment->element(track);
+            if ( element == 0)
+                  continue;
+            if ( element->type() != Element::Type::CHORD)
+                  continue;
+            {
+                  vector<int> harmonys;
+                  Chord *chord = static_cast<Chord*>(element);
+                  if ( calcFiguredBass(chord, harmonys) ) { // appends entries into harmonys corresponding musical harmony pitches
+                        for (int pitch : harmonys) {
+                              NoteVal *nval = new NoteVal(pitch);
+                              chord->score()->addNote(chord,*nval);
+                        }
+                  } else {
+                        //qDebug("no figured bass on chord on track %d", chord->track());
+                  }
+            }
+      }
+}
+      
+// populate the given harmonys vector with the pitches of the harmony notes implied by the figured base on the given chord
+bool FiguredBass::calcFiguredBass(Chord *chord, vector<int> & harmonys) {
+      Q_ASSERT(harmonys.size() == 0);
+      Segment *segment = chord->segment();
+      int octave = 0; // 0, or 12 if we want the harmony 1 octave up
+      FiguredBassItem::Modifier none = FiguredBassItem::Modifier::NONE;
+      if (find_if(segment->annotations().cbegin(),
+                  segment->annotations().cend(),
+                  [chord](Element *e) {
+                        return ( e->type() == Element::Type::FIGURED_BASS
+                                && e->track() == chord->track());
+                  }) == segment->annotations().cend())
+      {
+             return false; // no figured base to annotate
+      }
+      
+      if ( chord->notes().size() == 0 ) {
+            qDebug("no notes on chord");
+            return false; // no notes in the chord, not sure if this is possible?
+      }
+      if ( chord->notes().size() != 1 ) {
+            qDebug("measure #%d, cannot render Figured Base on a multi-note chord", segment->measure()->no());
+            return false;
+      }
+      Note *note = chord->notes()[0];
+
+      auto mod7 = [] (int n) {
+            n = n%7;
+            while ( n < 0)
+                  n += 7;
+            return n;
+      };
```

Comment:
I really hesitate responding in fear of starting an unnecessary flame war, but I would like to
explain my point of view.  Please accept my comments in the friendly spirit as they are intended.

With reference to the readability of the code I recently submitted.  I think it was a good idea to commit it
because it has allows people to examine the behavior, i.e., the algorithm I’m using for calculating the 
implied harmonies.  Doing so has allowed me to fix lots of problems with the algorithm.  However, please 
please keep in mind the code was quick and dirty, it was not intended to
be production worthy.  Please don’t be too upset it it is hard to read.  It was not optimized for readablity.

In my most recent version of the code, lots of the code people didn’t like actually went away not because
of style issues, but simply because it was making the wrong calculation and getting the wrong result.

That being said, It is strikingly interesting that people have difficulty with local functions. I never expected
it to be a controversial subject in 2015 and in C++11.  Of course local functions are useful, they are a 
feature of almost all  modern  programming languages.  C++ didn’t support them for a long time, but it
does now.  I don’t really see any reason to avoid them.  There’s nothing strange or scary about them.
The fact that C and C++ have not supported local function since the conception of the language in 
the 80s is not really a reason (in my opinion) to avoid them. In fact many of the features of C++11 
(this included) and the STL make the language much more accessible to non C++ hackers. 

I really respectfully reject the claim that local functions make the code more difficult to follow.   
On the contrary their proper use can make it EASIER to follow.   Their misuse can make things
more difficult.  The same can be said for the misuse of unsigned integers, operator overloading,
classes, templates, or even #include.   You can write bad code with whatever tools you have, 
that does not make the tools bad.

Local functions have many advantages.  I am happy to elaborate to anyone who is interested. That 
benefit is in no way limited to or limited by C++. Including

   reducing code size
   increasing runtime speed (because of aggressive inlining by the compiler)
   eliminating copy-paste bugs
   improving refactoring ability by reducing external dependencies.
   giving names to local operations
   improving readablity
   allowing code to express the intent
   improving maintainablity by keeping clutter out of global accessability
   improve IDE usability, by reducing numbers of global completions
   probably lots more

Yes, in a limited number of cases local functions CAN BE factored out making them globally accessible, but in 
the same sense that local constant variables could also be factored out to be visible globally.  In fact that is 
what a local function is, a local constant whose value is a function object

On the other hand, sometimes local functions CANNOT be made global as they may reference other variables in 
the scope in which they are defined.  This is what the leading [] means.  It tells the compiler which if any local 
variables are intentionally “captured”  by the function.  Usually local functions get inlined by the compiler (which 
is one large advantage of using them), but in the cases the compiler decides not to, the variables in the [] tell the 
compiler which variables must be passed in addition to the formal parameters and whether they are passed by 
value or by reference.

With reference to the auto keyword.  The variable IS NOT of type int, so an int declaration would be wrong.
The actual type of the variable is something complicated and uninteresting to the human, but very interesting to
the compiler.  The type delcares what type(s) of variable(s) the function accepts, whether those variables are passed by reference
or by value, and what the return type is. The auto keyword also makes sure the type is compatible with the STL functions like 
sort, find_if, find_if_not, find_first_of etc.



> On 25 May 2015, at 12:14, Maurizio M. Gavioli <notifications@github.com> wrote:
> 
> In libmscore/figuredbass.cpp <https://github.com/musescore/MuseScore/pull/2015#discussion_r30972851>:
> 
> > +      if ( chord->notes().size() == 0 ) {
> > +            qDebug("no notes on chord");
> > +            return false; // no notes in the chord, not sure if this is possible?
> > +      }
> > +      if ( chord->notes().size() != 1 ) {
> > +            qDebug("measure #%d, cannot render Figured Base on a multi-note chord", segment->measure()->no());
> > +            return false;
> > +      }
> > +      Note *note = chord->notes()[0];
> > +
> > +      auto mod7 = [] (int n) {
> > +            n = n%7;
> > +            while ( n < 0)
> > +                  n += 7;
> > +            return n;
> > +      };
> Partly for other duties, partly to think again and see if I am overlooking something, I took some time before answering. The question is:
> 
> Why using local functions at all? As they:
> 
> make the execution flow harder to understand
> lend to creating huge, complex 'parent' functions, which could be split into smaller pieces, easier to understand and maintain, with exactly the same result
> lend to code duplication (your local function mod7() is likely to be needed elsewhere in a programme dealing with musical scales, but the code would need to be repeated there)
> possibly make the life of IDE's and debuggers harder
> As @lasconic <https://github.com/lasconic> said, it is (mostly) a matter of taste, but I believe it is also a matter of economy: I fail to see any advantage and I feel several disadvantages in using local functions, unless they are absolutely necessary (a case, incidentally, I still have to meet). In addition, as lasconic also said, they are (almost?) never used in current MuseScore code base.
> 
> Very minor: why auto? can that function return anything else than an int? then, why obfuscating this piece of info under the auto return type?
> 
> Your code might be crystal clear and obvious to you -- as mine looks obvious to me --, but one of the tenets of open source code is that anybody else (including yourself six months from now) might be looking at your code in the future (to extend or simply maintain it) and making that anybody's life as easier as possible will help the life of the project.
> 
> I’m aware that my own code is far from obvious here and there (this is why I tend to exceed in comments) and regularly someone points these places out and starts asking embarassing questions; which is irritating, sure, but usually means that something is not adequate in those places.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/musescore/MuseScore/pull/2015/files#r30972851>.
>